### DISCUSSION TITLE
Macos Metal C++

### DISCUSSION THREAD

**Discussion by HgGamer** [ocornut/imgui#5544] [New Users Build/Link/Run issues ONLY!]
Hello!
I want to use imgui in a c++ only project, I tried the metal backend, but I'm not sure how to use it. 
I'm using Xcode and I have an m1 mac. 

My Code is the following:

```


#define NS_PRIVATE_IMPLEMENTATION
#define CA_PRIVATE_IMPLEMENTATION
#define MTL_PRIVATE_IMPLEMENTATION
#include <Foundation/Foundation.hpp>
#include <Metal/Metal.hpp>
#include <QuartzCore/QuartzCore.hpp>

#include "imgui/imgui.h"
#include "imgui/imgui_impl_metal.h"

int main(int argc, char* argv[])
{
    
    MTL::Device* pDevice = MTL::CreateSystemDefaultDevice();

    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO(); (void)io;
    ImGui::StyleColorsDark();
    ImGui_ImplMetal_Init(pDevice); // it breaks here
    
    pDevice->release();
  
    return 0;
}



```

And I'm getting the following error:

```
Undefined symbols for architecture arm64:
  "ImGui_ImplMetal_Init(MTL::Device*)", referenced from:
      _main in main.o
ld: symbol(s) not found for architecture arm64
```

**Comment by developer239** [comment]
@HgGamer From what I understand it is not possible to use metal because it only works in device having A7+ chip, simulator will not work for this. At least `CAMetalLayer` is not accessible on MacOS.

**Comment by developer239** [comment]
I have working simple example CMake + SDL + ImGui + SDLRenderer. It is almost 1AM. I will get back here by the end of the week with complete repository. Maybe I will also update `examples/example_sdl_sdlrenderer/main.cpp` because it seems that the event loop is not implemented correctly.

**Comment by developer239** [comment]
@HgGamer Not sure if this is still relevant but I am leaving it here for future reference. I am new to C++ but this is how I imagine the setup should look like. Ideally I would also like to pull SDL2 from submodule to avoid global dependencies. But there would be quite too many repositories that I would have to take care of so I will keep it as it is for now.

 https://github.com/developer239/cmake-sdl2-imgui

**Reply by HgGamer** [reply]
Thanks a lot, I will check it later. I will probably need to support metal API in my project later.

**Reply by developer239** [reply]
The `FindImgui` file that I created is tightly [coupled with SDL2 backend](https://github.com/developer239/cmake-modern-findimgui-sdl/blob/832e76903413ebce5933bb217e02186556550764/FindImgui.cmake). Usually I go against tight coupling but in this case I wanted to make opinionated setup that people can just clone and run. 

You will need to create your own cmake file and replace:
- [${IMGUI_LIBRARY}/backends/imgui_impl_sdl.cpp](https://github.com/developer239/cmake-modern-findimgui-sdl/blob/832e76903413ebce5933bb217e02186556550764/FindImgui.cmake)
- [${IMGUI_LIBRARY}/backends/imgui_impl_sdlrenderer.cpp](https://github.com/developer239/cmake-modern-findimgui-sdl/blob/832e76903413ebce5933bb217e02186556550764/FindImgui.cmake#L14)

files with metal related backend files.

And then I am not sure about [include directories](https://github.com/developer239/cmake-modern-findimgui-sdl/blob/832e76903413ebce5933bb217e02186556550764/FindImgui.cmake#L3). I wasn't able to make the imgui metal example work because `CAMetalLayer` was not available on my machine. It seems it is something that is related to mobile devices.

Other than that you just need to put it in `externals` and load it as any other CMake library. You can use this repository as a reference:

https://github.com/LeeTeng2001/metal-cpp-cmake

**Reply by developer239** [reply]
@HgGamer Btw I had a typo in the CMake file so you should update the fork. 🫠
